source/common/common/perf_annotation.h

```diff
@@ -0,0 +1,184 @@
+ * that an uncontended mutex lock has vanishingly small cost. In the future we may try
+ * to make this system thread-unsafe if mutext contention disturbs the metrics.
+ */
+#define PERF_THREAD_SAFE true
+
+namespace Envoy {
+
+/**
+ * Defines a context for collecting performance data. Note that this class is
+ * fully declared and defined even if ENVOY_PERF_AUTOMATION is off. We depend on
+ * the macros to disable performance collection for production.
+ */
+class PerfAnnotationContext {
+public:
+  /**
+   * Records time consumed by a category and description, which are shown as separate
+   * columns in the generated output table.
+   *
+   * @param duration std::chrono::nanoseconds the duration.
+   * @param category absl::string_view the name of a category for the recording.
+   * @param category absl::string_view the name of description for the recording.
+   *
+   */
+  void record(std::chrono::nanoseconds duration, absl::string_view category,
+              absl::string_view description);
+
+  /**
+   * Renders the aggregated statistics as a string.
+   * @return std::string the performance data as a formatted string.
+   */
+  static std::string toString();
+
+  /**
+   * Dumps aggregated statistics (if any) to stdout.
+   */
+  static void dump();
+
+  /**
+   * Thread-safe lazy-initialization of a PerfAnnotationContext on first use.
+   * @return PerfAnnotationContext* the context.
+   */
+  static PerfAnnotationContext* getOrCreate();
+
+  /**
+   * Clears out all aggregated statistics.
+   */
+  static void clear();
+
+private:
+  /**
+   * PerfAnnotationContext construction should be done via getOrCreate().
+   */
+  PerfAnnotationContext();
+
+  using CategoryDescription = std::pair<std::string, std::string>;
+
+  struct DurationStats {
+    std::chrono::nanoseconds total_{0};
+    std::chrono::nanoseconds min_{0};
+    std::chrono::nanoseconds max_{0};
+    WelfordStandardDeviation stddev_;
+  };
+
+  struct Hash {
+    size_t operator()(const CategoryDescription& a) const {
+      return std::hash<std::string>()(a.first) + 13 * std::hash<std::string>()(a.second);
+    }
+  };
+
+  using DurationStatsMap = std::unordered_map<CategoryDescription, DurationStats, Hash>;
+
+  DurationStatsMap duration_stats_map_; // Maps {category, description} to DurationStats.
+#if PERF_THREAD_SAFE
+  std::mutex mutex_;
+#endif
+};
+
+/**
+ * Represents an operation for reporting timing to the perf system. Usage:
+ *
+ * f() {
+ *   PerfOperation perf_op;
+ *   computeIntensiveWork();
+ *   perf_op.record("category", "description");
+ * }
+ */
+class PerfOperation {
+public:
+  PerfOperation();
+
+  /**
+   * Report an event relative to the operation in progress. Note report can be called
+   * multiple times on a single PerfOperation, with distinct category/description combinations.
+   * @param category absl::string_view the name of a category for the recording.
```

Comment:
FYI, we don't typically include types in `@param`. I know, I know, none of this makes any sense other than from a consistency argument perspective.

Reply:
my bad; this is explicit in STYLE.md.  Done.